### DISCUSSION TITLE
LDT unable to read igbp.bin

### DISCUSSION THREAD

**Discussion by kbergaoui** [NASA-LIS/LISF#872] [Q&A] [answered]
### Bug Description
I compiled LISF-support-lisf-public-7.3 using oneAPI 2021 version under Ubuntu 20.04 where LDT is compiled in serial mode.
When I execute ./LDT ldt.config I obtain the following error:
Image              PC                Routine            Line        Source             
LDT                0000000000AC907A  Unknown               Unknown  Unknown
libpthread-2.31.s  00007F14FB33F3C0  Unknown               Unknown  Unknown
LDT                00000000008CD7C2  read_modisnative_         165  read_MODISNative_lc.F90
LDT                00000000004F84E5  ldt_lmlcmod_mp_lm         349  LDT_LMLCMod.F90
LDT                00000000005F6A76  ldt_paramprocmod_         268  LDT_paramProcMod.F90
LDT                000000000059C495  ldt_init_lsmparam          18  LDT_init_LSMparamproc.F90
LDT                000000000064CE10  MAIN__                     39  LDTmain.F90
LDT                000000000041D7E2  Unknown               Unknown  Unknown
libc-2.31.so       00007F14FB15D0B3  __libc_start_main     Unknown  Unknown
LDT                000000000041D6EE  Unknown               Unknown  Unknown

the log file is attached:
[ldtlog.0000.txt](https://github.com/NASA-LIS/LISF/files/6918242/ldtlog.0000.txt)
the namelist file is attached:
[ldt.config.txt](https://github.com/NASA-LIS/LISF/files/6918248/ldt.config.txt)
the installation script (including all the external libraries) is attached:
[install_lis.txt](https://github.com/NASA-LIS/LISF/files/6918290/install_lis.txt)

### Steps to Replicate

./LDT ldt.config

**Comment by bmcandr** [comment]
Hi @kbergaoui, I have converted this Issue to a Discussion until we can confirm that you have identified a bug.

**Comment by emkemp** [comment]
Hi @kbergaoui:

Can you try running ./configure again with Optimization level to -2, run ./compile again, and then rerun your case?

The -2 flag can help provide more informative error messages in the stacktrace.

**Reply by dmocko** [reply]
Hi,

I might also suggest changing this config entry:

Landcover fill option:          neighbor

David

**Reply by kbergaoui** [reply]
Hello,

I recompiled LIS and LDT with the -2 option and used the neighbor entry as follows:
Landcover fill option:   neighbor    
Landcover fill radius:     3.        
Landcover fill value:      5.        

I got the same error:
Image              PC                Routine            Line        Source             
LDT                00000000028D044A  Unknown               Unknown  Unknown
libpthread-2.31.s  00007F39670E63C0  Unknown               Unknown  Unknown
LDT                0000000001F971CE  read_modisnative_          94  read_MODISNative_lc.F90
LDT                0000000000D231A4  Unknown               Unknown  Unknown
LDT                0000000000925844  ldt_lmlcmod_mp_lm         355  LDT_LMLCMod.F90
LDT                0000000000E1D1C8  ldt_paramprocmod_         274  LDT_paramProcMod.F90
LDT                0000000000CEB030  ldt_init_lsmparam          24  LDT_init_LSMparamproc.F90
LDT                0000000000F44C27  Unknown               Unknown  Unknown
LDT                0000000001101622  MAIN__                     45  LDTmain.F90
LDT                000000000041D7E2  Unknown               Unknown  Unknown
libc-2.31.so       00007F3966F040B3  __libc_start_main     Unknown  Unknown
LDT                000000000041D6EE  Unknown               Unknown  Unknown


and please find attached the log file error
[ldtlog.0000.txt](https://github.com/NASA-LIS/LISF/files/6919455/ldtlog.0000.txt)

**Reply by kbergaoui** [reply]
I've reinstalled LIS from the master release 7.4 (the latest one including all the updates). (Please find attached the configure.ldt and the compilation output).
I've got exactly the same error!
[compile.out.txt](https://github.com/NASA-LIS/LISF/files/6923091/compile.out.txt)
[ldtlog.0000.txt](https://github.com/NASA-LIS/LISF/files/6923092/ldtlog.0000.txt)
[configure.ldt.txt](https://github.com/NASA-LIS/LISF/files/6923097/configure.ldt.txt)

**Reply by emkemp** [reply]
Hi @kbergaoui 

Okay, so a few findings.  First, you are correctly setting the endianness to "big", so we can scratch that off.  Second, the error seems to occur when the program inquires about whether the landcover file exists:

!- Check if land cover file exists:
   inquire( file=trim(LDT_rc%vfile(n)), exist=file_exists )  !!!<<<< This is line 94, where it crashes
   if(.not. file_exists) then
      write(LDT_logunit,*)"[ERR] The landcover map: ",trim(LDT_rc%vfile(n))," does not exist."
      write(LDT_logunit,*)" Program stopping ..."
      call LDT_endrun
   endif

I'm guessing the LDT_rc%vfile(n) variable has garbled data.  I suggest modifying the code with write statements in two places.

First, in ldt/params/landcover/read_MODISNative_lc.F90, before line 94, add this line:

write(LDT_logunit,*)'TEST: n = ', n  ! <<< NEW
write(LDT_logunit,*)'TEST: LDT_rc%vfile(n) =  ', trim(LDT_rc%vfile(n)) ! <<< NEW

Second, in core/LDT_LMLCMod.F90, before line 191 (and *in* the do loop) add similar write statements.  So:

    call ESMF_ConfigFindLabel(LDT_config,"Landcover file:",rc=rc)
    do n=1,LDT_rc%nnest
       call ESMF_ConfigGetAttribute(LDT_config,LDT_rc%vfile(n),rc=rc)
       call LDT_verify(rc,'Landcover file: not specified')
       write(LDT_logunit,*)'TEST: n = ', n  ! <<< NEW
       write(LDT_logunit,*)'TEST: LDT_rc%vfile(n) =  ', trim(LDT_rc%vfile(n))  ! <<< NEW
    enddo

Then, recompile LDT.

I'm not sure why the LDT_rc%vfile variable would be corrupted, since your ldt.config indicates you are only using 1 nest.  But this is my best guess right now.

**Reply by kbergaoui** [reply]
I added the writing statements in both routines. In the log file (attached) I've got just the writing in LDT_LMLCMod.F90 and not the one in read_MODISNative_lc.F90.
However the writing was correct with n=1 and the name and location of the vfile are correct.
[ldtlog.0000.txt](https://github.com/NASA-LIS/LISF/files/6928794/ldtlog.0000.txt)
[ldtlog.0000.txt](https://github.com/NASA-LIS/LISF/files/6928797/ldtlog.0000.txt)

**Reply by emkemp** [reply]
Hi @kbergaoui 

I went ahead and set up your LDT case (except for the irrigation-fraction map) on the NASA Discover supercomputer, adjusting the paths appropriately.  I cannot reproduce your problem, though I note that I'm using older Intel compilers (19.1).

However, I do note that your stacktrace doesn't show a specific error message.  Normally when an ifort-compiled program fails, there is a message above the stacktrace like:

forrtl: error (65): floating invalid

or something similar.  Do you see such an error message?

Since my job runs on Discover, another possibility is your computer is running out of memory.  I note that ldt/params/landcover/read_MODISNative_lc.F90 automatically allocates a number of local arrays when that subroutine starts.  Can you update your run environment by running:

ulimit -c unlimited
ulimit -m unlimited
ulimit -s unlimited
ulimit -S unlimited

(This syntax assumes you are running the bash shell.)

**Reply by kbergaoui** [reply]
Hi Eric
The error message I have from intel is:
forrtl: severe (174): SIGSEGV, segmentation fault occurred

The issue is not related to memory because during the run I visualize the memory use and it doesn't exceed 2 Giga (The available memory is 128 Giga)

Please kindly Note that when I put another write (logunit,*) statement in the routine 
LDT_LMLCMod.F90 (The modified routine is attached) just after the line 220 I obtain a different error related to the change of n ( n= 2) (Please see attached logfile)
[LDT_LMLCMod.F90.txt](https://github.com/NASA-LIS/LISF/files/6936796/LDT_LMLCMod.F90.txt)
[ldtlog.0000.txt](https://github.com/NASA-LIS/LISF/files/6936812/ldtlog.0000.txt)


Unfortunately I am using intel 2021 (oneAPI) because I can't access the free version of intel 2019.

**Reply by emkemp** [reply]
Hi @kbergaoui:

The new error you get with n=2 is just a result of putting those write statements after the do loop rather than in it.  When you run:

do n=1, LDT_rc%nnest
...
enddo

the program will not exit the loop until n > LDT_rc%nnest.  Since LDT_rc%nnest = 1, the loop is exited when n=2.  Since nothing was assigned to LDT_rc%vfile(2) (indeed, I think that array was allocated to size 1), the program fails at that point.

Important question:  Is the SIGSEGV error message from the original crash, or from the run you just did with the moved write statements?

SIGSEGV indicates the program is either accessing memory that doesn't belong to it, or is accessing memory in a way that is not permitted (e.g., writing to read-only memory).

**Reply by kbergaoui** [reply]
Ah ok.
the SIGSEGV error message comes from the original crash

**Reply by emkemp** [reply]
Also: SIGSEGV can occur if the program exceeds the amount of "stack memory" on the computer, which is usually much smaller than the total RAM.  The read_MODISNative_lc subroutine has a lot of arrays that are automatically allocated, and by default these allocations occur on the stack:

   real      :: go1(LDT_rc%lnc(n)*LDT_rc%lnr(n))            ! Output lis 1d grid
   logical*1 :: lo1(LDT_rc%lnc(n)*LDT_rc%lnr(n))            ! Output logical mask (to match go)
   real      :: go2(LDT_rc%lnc(n)*LDT_rc%lnr(n), LDT_rc%nt) ! Output lis 1d grid
   logical*1 :: lo2(LDT_rc%lnc(n)*LDT_rc%lnr(n), LDT_rc%nt) ! Output logical mask (to match go)

*If* that is the problem, there are two solutions:

The first is to set ulimit -s unlimited (assuming you are using the bash shell).  That raises the stack size limit to a much higher amount (either the total RAM or some system-wide limit set by the system administrator.)

The second is to edit the configure.ldt file and add this compiler flag to FFLAGS and FFLAGS77:

-heap-arrays

This is an ifort-specific flag that forces automatic array allocation on the "heap" (a larger subset of RAM).  It works with Intel 19, but I don't know if it is available with Intel 2021 (you'll need to check your documentation).  Note that there can be a speed penalty with this flag.

**Reply by emkemp** [reply]
BTW, if you are using the bash shell, you can use 'ulimit -a' to list all of the current settings, including maximum stack size.

**Reply by kbergaoui** [reply]
Many thanks Eric. The ulimit -s unlimited   command fixed the problem!
The error I'm getting now is the following:
Warning 1: /home/erdms/lis/LS_PARAMETERS/soil_parms/ISRIC/v2017/TEXMHT_M_sl1_250m.tif: /home/erdms/lis/LS_PARAMETERS/soil_parms/ISRIC/v2017/TEXMHT_M_sl1_250m.tif:AdobeDeflate compression support is not configured
ERROR 1: /home/erdms/lis/LS_PARAMETERS/soil_parms/ISRIC/v2017/TEXMHT_M_sl1_250m.tif: Cannot open TIFF file due to missing codec.

and this is probably related to the tif external library I've installed. I'll re-install it with the latest version and with this AdobeDeflate compression support

Many thanks again for your precious support.

**Reply by emkemp** [reply]
Excellent!

Since igbp.bin can be read, I will close this discussion.  Further problems (if any) can be handled with separate discussions.

**Reply by emkemp** [reply]
Actually I can't close the discussion.  @kbergaoui, do you see an option on your screen to mark the discussion as "Answered"?

**Reply by emkemp** [reply]
Disregard, I found the answer checkmark.

**Comment by emkemp** [comment, marked as answer]
The cause was insufficient stack space on the local machine.  Setting 'ulimit -s unlimited' allowed LDT to read the file.
